### PR TITLE
Generate header with version #defines for LibTorch

### LINKED ISSUE #44365: Add ability to determine libtorch version from a c++ API or preprocessor macro.
## 🚀 Feature
I want to be able to determine the version of libtorch in my c++ binary in cases where I don't have access to the python library.

Something like:
```
#define TORCH_VERSION_MAJOR=1
#define TORCH_VERSION_MINOR=6
#define TORCH_VERSION_PATCH=0
```

## Motivation

This is useful for 
- Turning off torch features depending on the version of torch we are running
- Debugging which version of torch is installed in production where different versions may be running.

## Pitch

1) I'm a developer trying to upgrade the API, but a backwards incompatible API change requires me to refactor code before upgrading.  I need to roll this out so I want to only enable the new features if I build my application for the new version of torch.
2) I'm a site reliability engineer and I am rolling out a the server deploy with a new version of libtorch.  There is a bug that is claimed to be fixed in the newer version of torch and I want to verify that the new version was deployed when I validate this issue is resolved.

## Alternatives

Currently I am pulling the version from the strings in libtorch_cpu.so, but this didn't get updated in v1.6.0.  This is an unreliable hack solution that I would like to remove.
```
strings libtorch_cpu.so | egrep -A1 pytorch_version
pytorch_version
1.5.0
```

pytorch has the version number baked into the pip install and in the torch.__version__.  For my purposes I don't have access to either of these as we are not using python in our production environment and the shared libraries are built separate from the ones in the pip wheel.

## Additional context

I recently upgraded to libtorch version 1.6.0 to resolve a segfault on destruction of `DeviceThreadHandlePool` which seemed related to: https://github.com/pytorch/pytorch/pull/36416.  After upgrading I am still seeing some segfaults on destruction, but was worried that I might not have deployed the new version of libtorch.   By looking at other strings in that library I believe I am at v1.6.0, but I wasted some time on the red herring of the 1.5.0 version in the strings of libtorch_cpu.so and would like to avoid that in the future.

cc @yf225 @glaringlee

### PR DESCRIPTION
**Pull request by jbschlosser**
Uses cmake's `configure_file()` macro to generate a new `torch/csrc/api/include/torch/version.h` header with `TORCH_VERSION_{MAJOR,MINOR,PATCH}` \#defines from an input file `torch/csrc/api/include/torch/version.h.in`.

For Bazel builds, this is accomplished with `genrule()` and a python generation script.

Fixes #44365

Screen Shot 2021-01-05 at 3 19 24 PM

### CODE REVIEW

File: caffe2/CMakeLists.txt
Comment (malfet) [inline caffe2/CMakeLists.txt:341]: You'll probably need to add `header_template_rule` to BUILD.bazel similar to 
https://github.com/pytorch/pytorch/blob/12ee7b61e75caa0c57b53ee1627e3075686d2510/BUILD.bazel#L41-L48

Comment (glaringlee) [reply caffe2/CMakeLists.txt:341]: > You'll probably need to add `header_template_rule` to BUILD.bazel similar to
> https://github.com/pytorch/pytorch/blob/12ee7b61e75caa0c57b53ee1627e3075686d2510/BUILD.bazel#L41-L48

@malfet 
It seems to me that we did not setup the version macros in BUILD.bazel (correct me if I am wrong)
I find this https://github.com/pytorch/pytorch/blob/master/BUILD.bazel#L786, but is it hard coded to me.

**Comment by glaringlee** [comment]
If the bzl is not experimental and we need to setup the version correctly, given that bazel can not read file directly, we might need to create action to read version.txt and setup the TORCH_VERSION_XXXX just like what we do here, https://github.com/pytorch/pytorch/blob/9529ae377629f4253af0720cb01f09db6fcad4c9/CMakeLists.txt#L463

File: caffe2/CMakeLists.txt
Comment (malfet) [reply caffe2/CMakeLists.txt:341]: @glaringlee we do not ship any binaries from bazel build, so version numbers can be stale.

Comment (glaringlee) [reply caffe2/CMakeLists.txt:341]: @malfet thanks for comment.
@jbschlosser per @malfet's comment above, it is fine that we just use a hard coded version for TORCH_VERSION_XXXX in bzl, but if you are in the middle of looking at bzl action to parse version.txt and it is not too time consuming, it is also great to let bzl support versioning. You can do either way, let me know when you think this is ready.

Comment (jbschlosser) [reply caffe2/CMakeLists.txt:341]: Great- thanks for looking into it! I have it working for now hard-coded, but I want to attempt the proper solution. Will let you know when I'm ready.

Comment (glaringlee) [reply caffe2/CMakeLists.txt:341]: sure, thank you!

Comment (jbschlosser) [reply caffe2/CMakeLists.txt:341]: @glaringlee Well, I now have the version header being generated from `version.txt` in Bazel. Since Bazel cannot read file contents at analysis time in a way that affects the build (see [here](https://stackoverflow.com/questions/56007307/read-local-file-content-in-bazel-build-system)), I could not use `header_template_rule` and had to switch to a `genrule` that launches a python script that does manual template expansion. It works but it's ugly. Perhaps it's best to keep the hard-coded version for now and solve this properly later on.

Comment (glaringlee) [reply caffe2/CMakeLists.txt:341]: @malfet This PR now uses `genrule` to parse version.txt in bazel build to support torch version number in c++.  The code introduced a helper .py script to parse version.txt and generate the version header file, it seems fine to me. Can you advise?

File: tools/setup_helpers/gen_version_header.py
Comment (malfet) [inline tools/setup_helpers/gen_version_header.py:26]: [PEP-8](https://www.python.org/dev/peps/pep-0008/#blank-lines) suggest 2 blank lines between functions

**Review by malfet — APPROVED** [review]
Thank you for the thorough fix!

File: tools/setup_helpers/gen_version_header.py
Comment (glaringlee) [reply tools/setup_helpers/gen_version_header.py:26]: @jbschlosser fix this, and we are good to land

**Review by glaringlee — APPROVED** [review]
Approved, but need to fix the bazel test failure before landing.
I think we need to add the generated header here:
https://github.com/pytorch/pytorch/pull/50073/files#diff-7fc57714ef13c3325ce2a1130202edced92fcccc0c6db34a72f7b57f60d552a3R1903

**Comment by glaringlee** [comment]
@jbschlosser 
It seems this passed the bazel build, but bazel test failed.

**Comment by jbschlosser** [comment]
> Approved, but need to fix the bazel test failure before landing.
> I think we need to add the generated header here:
> https://github.com/pytorch/pytorch/pull/50073/files#diff-7fc57714ef13c3325ce2a1130202edced92fcccc0c6db34a72f7b57f60d552a3R1903

Ah, you're right. I thought the glob on `torch/csrc/**/*.h` would pick up the generated header, but it makes sense that it has to be explicitly included for the generation rule to trigger.

**Review by glaringlee — APPROVED** [review]
This is good now.

**Pull request closed without merging**
